Optimizing creation and addition of iobref

rkothiya opened the pull request:
Fixes: #1620

Change-Id: I68cf67b054fc5aea76d936bf5096e57509146038
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
Optimizing creation and addition of iobref

Fixes: #1620

Change-Id: I68cf67b054fc5aea76d936bf5096e57509146038
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

mykaul commented on cli/src/cli-quotad-client.c:
```diff
@@ -34,16 +34,14 @@ cli_quotad_submit_request(void *req, call_frame_t *frame, rpc_clnt_prog_t *prog,
             }
 
             new_iobref = 1;
         }
```
You are missing an else part - if there's already an iobref, you need to have the iobref_add() function.

rkothiya replied:
Done

rkothiya commented:
/run regression

Commit:
Optimizing creation and addition of iobref

Added the missing else section.
If there's already an iobref then used it in iobref_add.

Fixes: #1620

Change-Id: Id5cda83f50d6b8230748a5ee8b09bb1843569dc3
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/run regression

rkothiya commented:
Tested georep-basic-rsync-ec.t on Fedora32, its failing even without this patch : 

~~~
Test Summary Report
-------------------
/home/rinku/Code/Gluster/glusterfs/tests/00-geo-rep/georep-basic-rsync-ec.t (Wstat: 0 Tests: 82 Failed: 36)
  Failed tests:  24-31, 34-35, 39-46, 49-51, 54-59, 62-64
                67, 72-76
Files=1, Tests=82, 4385 wallclock secs ( 1.73 usr  1.12 sys + 118.99 cusr 208.98 csys = 330.82 CPU)
Result: FAIL
~~~

rkothiya commented:
/run regression

mykaul commented on api/src/glfs-fops.c:
```diff
@@ -6071,16 +6071,8 @@ glfs_anonymous_pwritev(struct glfs *fs, struct glfs_object *object,
-    ret = iobref_add(iobref, iobuf);
-    if (ret) {
         iobuf_unref(iobuf);
         iobref_unref(iobref);
```
I'm not sure this is needed.

rkothiya replied:
This is not needed, hence removed it.

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -1119,16 +1135,8 @@ iobuf_copy(struct iobuf_pool *iobuf_pool, const struct iovec *iovec_src,
-    ret = iobref_add(*iobref, *iobuf);
-    if (ret) {
         iobuf_unref(*iobuf);
         iobref_unref(*iobref);
```
Is this needed?

rkothiya replied:
This is not needed, hence removed it.

mykaul commented on rpc/rpc-lib/src/rpcsvc.c:
```diff
@@ -1302,16 +1300,14 @@ rpcsvc_callback_submit(rpcsvc_t *rpc, rpc_transport_t *trans,
             goto out;
         }
         new_iobref = 1;
     }
```
Missing else part?

rkothiya replied:
Done

mykaul commented on rpc/rpc-transport/socket/src/socket.c:
```diff
@@ -1799,19 +1799,15 @@ __socket_read_accepted_successful_reply(rpc_transport_t *this)
                         iobuf_unref(iobuf);
                         goto out;
                     }
                 }
```
Missing an else part?

rkothiya replied:
Done

mykaul reviewed (changes requested):
You are missing in several code paths the 'else' part - if the iobref already exists, all you need to do is the add part.

Commit:
Optimizing creation and addition of iobref

Added the missing else section. If there's already
an iobref then used it in iobref_add.

Removed the unnecessary calls to iobref_unref()

Fixes: #1620

Change-Id: Ief2e61f6262826333e22792b7dd4f3149d0f411f
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/georep-basic-rsync-ec.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/georep-basic-rsync-ec.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/1289/

rkothiya commented:
As the regression test failed due to ./tests/00-geo-rep/georep-basic-rsync-ec.t, I tried running this test without the patch and noticed that even without the patch this test is failing. 

~~~
not ok  65 [     90/ 120200] < 158> 'x0 arequal_checksum /mnt/glusterfs/0/logrotate /mnt/glusterfs/1/logrotate' -> 'Got "x21" instead of "x0"'
ok  66 [     46/     61] < 161> '0 create_rename_ok /mnt/glusterfs/1/create_rename_test_file'
not ok  67 [     45/ 120126] < 164> '0 hardlink_rename_ok /mnt/glusterfs/1/hardlink_rename_test_file' -> 'Got "1" instead of "0"'
ok  68 [     67/   4203] < 167> 'gluster --mode=script --wignore volume geo-replication primary 127.0.0.1::secondary stop'
ok  69 [     59/    283] < 170> 'create_symlink_rename_mkdir_data'
ok  70 [     57/    198] < 173> 'create_hardlink_rename_data'
ok  71 [     80/   3242] < 179> 'gluster --mode=script --wignore volume geo-replication primary 127.0.0.1::secondary start'
not ok  72 [     70/ 121985] < 182> '2 check_status_num_rows Active' -> 'Got "0" instead of "2"'
not ok  73 [     57/ 121179] < 183> '4 check_status_num_rows Passive' -> 'Got "0" instead of "4"'
not ok  74 [     93/ 120034] < 187> '0 hardlink_rename_ok /mnt/glusterfs/1/hardlink_rename_test_file' -> 'Got "1" instead of "0"'
not ok  75 [    133/ 120046] < 190> '0 verify_symlink_rename_mkdir_data /mnt/glusterfs/1/symlink_test1' -> 'Got "1" instead of "0"'
not ok  76 [     41/ 120100] < 193> '0 verify_hardlink_rename_data /mnt/glusterfs/1' -> 'Got "1" instead of "0"'
ok  77 [     88/   1561] < 199> 'gluster volume geo-rep primary 127.0.0.1::secondary config rsync-options --whole-file'
ok  78 [     16/     35] < 200> 'echo sampledata > /mnt/glusterfs/0/rsync_option_test_file'
not ok  79 [     13/ 120122] < 208> 'x0 arequal_checksum /mnt/glusterfs/0 /mnt/glusterfs/1' -> 'Got "x28" instead of "x0"'
ok  80 [     43/   1892] < 211> 'gluster --mode=script --wignore volume geo-replication primary 127.0.0.1::secondary stop'
ok  81 [     28/   1915] < 214> 'gluster --mode=script --wignore volume geo-replication primary 127.0.0.1::secondary delete'
ok  82 [     39/      3] < 217> 'rm ./tests/00-geo-rep/../utils/arequal-checksum'
Failed 38/82 subtests 

Test Summary Report
-------------------
./tests/00-geo-rep/georep-basic-rsync-ec.t (Wstat: 0 Tests: 82 Failed: 38)
  Failed tests:  24-31, 34-35, 39-46, 49-51, 54-59, 62-65
                67, 72-76, 79
Files=1, Tests=82, 4612 wallclock secs ( 1.46 usr  1.08 sys + 128.66 cusr 217.37 csys = 348.57 CPU)
Result: FAIL
~~~
